protocols/imap/src/main/java/org/apache/james/imap/api/ConnectionCheckFactory.java

```diff
@@ -0,0 +1,56 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.imap.api;
+
+import java.util.Optional;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+import org.apache.james.utils.ClassName;
+import org.apache.james.utils.GuiceGenericLoader;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.collect.ImmutableSet;
+
+@FunctionalInterface
+public interface ConnectionCheckFactory {
+    Set<ConnectionCheck> create(ImapConfiguration imapConfiguration);
+
+    public static class Impl implements ConnectionCheckFactory {
+        private final GuiceGenericLoader loader;
+
+        @Inject
+        public Impl(GuiceGenericLoader loader) {
+            this.loader = loader;
+        }
+
+        @Override
+        public Set<ConnectionCheck> create(ImapConfiguration imapConfiguration) {
+            return Optional.ofNullable(imapConfiguration.getConnectionChecks())
+                .orElse(ImmutableSet.of())
+                .stream()
+                .map(ClassName::new)
+                .map(Throwing.function(loader::instantiate))
+                .map(ConnectionCheck.class::cast)
+                .collect(ImmutableSet.toImmutableSet());
+        }
+    }
```

Comment:
Remove this impl class to another module that handles guice loading/binding stuff e.g. at the same module with `server/container/guice/protocols/imap/src/main/java/org/apache/james/modules/protocols/IMAPServerModule.java`? So we can remove the `guice-utils` out of `protocols/imap`.

The idea is `protocols/imap` does not need to care about guice loading-binding duty.

Reply:
So can we move `Impl` class into another module? E.g: `server/container/util/src/main/java/org/apache/james/util`

Reply:
wherever appropriate (I can sit down with u tmr), but for sure not in `protocols/imap` as ATM.